### PR TITLE
Json

### PR DESCRIPTION
**Pull request by Fhrozen**
Fix bugs for decoding
Convert2json & mergejson are now 1 file > mergejson

Tested with AN4 (The WER increased by 4, rather there is not modification of training files, it seems strange)

### CODE REVIEW

**Comment by sw005320** [comment]
@oadams, we're now simplifying `data2json.sh`. So, please monitor this PR.

**Comment by Fhrozen** [comment]
@sw005320 Could you please test with babel, or another example which uses especial characters?, I am little worry about that part, because the split_json print to file directly rather than use a stdout from the bash console.

**Comment by sw005320** [comment]
OK. I'm working on that.

**Comment by sw005320** [comment]
The order of utterances in `data.json` seems to be changed, which may affect the performance. Please check it.

File: egs/tedlium/asr1/run.sh [hunk 1]
```diff
@@ -152,6 +152,11 @@ if [ ${stage} -le 2 ]; then
          data/${train_set} ${dict} > ${feat_tr_dir}/data.json
     data2json.sh --feat ${feat_dt_dir}/feats.scp \
          data/${train_dev} ${dict} > ${feat_dt_dir}/data.json
+    for rtask in ${recog_set}; do
+        feat_recog_dir=${dumpdir}/${rtask}/delta${do_delta}
+        data2json.sh --feat ${feat_recog_dir}/feats.scp \
+            --nlsyms ${nlsyms} data/${rtask} ${dict} > ${feat_recog_dir}/data.json
```
Comment (sw005320) [inline egs/tedlium/asr1/run.sh]: This does not need `--nlsyms ${nlsyms}`

File: egs/voxforge/asr1/run.sh [hunk 2]
```diff
@@ -163,6 +163,11 @@ if [ ${stage} -le 2 ]; then
          data/tr_${lang} ${dict} > ${feat_tr_dir}/data.json
     data2json.sh --lang ${lang} --feat ${feat_dt_dir}/feats.scp \
          data/dt_${lang} ${dict} > ${feat_dt_dir}/data.json
+    for rtask in ${recog_set}; do
+        feat_recog_dir=${dumpdir}/${rtask}/delta${do_delta}
+        data2json.sh --feat ${feat_recog_dir}/feats.scp \
+            --nlsyms ${nlsyms} data/${rtask} ${dict} > ${feat_recog_dir}/data.json
```
Comment (sw005320) [inline egs/voxforge/asr1/run.sh]: This does not need `--nlsyms ${nlsyms}`

File: src/utils/splitjson.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+    filename = os.path.basename(args.json).split('.')[0]
+    dirname = os.path.dirname(args.json)
+    dirname = '{}/split{}utt'.format(dirname, args.parts)
+
+    if not os.path.exists(dirname):
+        os.makedirs(dirname)
+
+    for i in range (args.parts):
+        new_dic = {}
```
Comment (Fhrozen) [inline src/utils/splitjson.py]: Still reviewing this part.

File: src/utils/splitjson.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        for id in ids:
```
Comment (Fhrozen) [inline src/utils/splitjson.py]: still reviewing this part

**Comment by Fhrozen** [comment]
Fast test on AN4 with CPU & pytorch.

**Comment by sw005320** [comment]
LGTM. Please let me know how it affects the performance. If it is small, I can merge it.

**Comment by Fhrozen** [comment]
AN4 test:
`./run.sh --stage 3 --ngpu 1 --tag debug_alpha0.5_pt`
- Espnet Master:
```
|   SPKR     |   # Snt     # Wrd   |   Corr       Sub        Del        Ins        Err      S.Err   |
|   Sum/Avg  |    100       1915   |   77.4       7.4       15.2        0.6       23.1       78.0   |
|  SPKR     |   # Snt    # Wrd   |  Corr        Sub       Del       Ins        Err     S.Err   |
|  Sum/Avg  |    130      2565   |  84.3        5.6      10.1       1.1       16.8      68.5   |
```
- New JSON:
```
|   SPKR     |   # Snt     # Wrd   |   Corr       Sub        Del        Ins        Err      S.Err   |
|   Sum/Avg  |    100       1915   |   77.9       7.8       14.4        0.7       22.9       79.0   |
|  SPKR     |   # Snt    # Wrd   |  Corr        Sub       Del       Ins        Err     S.Err   |
|  Sum/Avg  |    130      2565   |  85.2        5.7       9.0       1.2       16.0      67.7   |

```
`./run.sh --stage 3 --ngpu 1 --tag debug_alpha1.0_pt --ctc_weight 1.0 --mtlalpha 1.0 --recog_model loss.best `
- Espnet Master:
```
|   SPKR     |   # Snt     # Wrd   |   Corr        Sub        Del        Ins        Err      S.Err   |
|   Sum/Avg  |    100       1915   |   82.6        5.0       12.5        1.4       18.9       80.0   |
|  SPKR     |   # Snt    # Wrd   |   Corr       Sub        Del       Ins        Err     S.Err   |
|  Sum/Avg  |    130      2565   |   90.6       2.8        6.6       1.5       10.9      64.6   |
```
- New JSON:
```
|   SPKR     |   # Snt     # Wrd   |   Corr        Sub        Del        Ins        Err      S.Err   |
|   Sum/Avg  |    100       1915   |   81.9        5.2       12.9        1.0       19.1       77.0   |
|  SPKR     |   # Snt    # Wrd   |   Corr       Sub        Del       Ins        Err     S.Err   |
|  Sum/Avg  |    130      2565   |   89.4       3.4        7.3       1.4       12.0      73.1   |
```

**Comment by sw005320** [comment]
OK. This looks reasonable. Can you also test the other task? My grid-engine slots are full, and I cannot test it.

**Comment by Fhrozen** [comment]
Voxforge with default parameters (multiGPU chainer):
`grep -e Avg -e SPK -m 2 exp/tr_it_vggblstmp_e4_subsample1_2_2_1_1_unit320_proj320_d1_unit300_location_aconvc10_aconvf100_mtlalpha0.5_adadelta_bs30_mli800_mlo150//decode_*/result.txt`

JSON:

```
exp/tr_it_vggblstmp_e4_subsample1_2_2_1_1_unit320_proj320_d1_unit300_location_aconvc10_aconvf100_mtlalpha0.5_adadelta_bs30_mli800_mlo150//decode_dt_it_beam20_eacc.best_p0_len0.0-0.0_ctcw0.3/result.txt:|       SPKR                         |       # Snt               # Wrd        |       Corr                 Sub                Del                 Ins                 Err               S.Err        |
exp/tr_it_vggblstmp_e4_subsample1_2_2_1_1_unit320_proj320_d1_unit300_location_aconvc10_aconvf100_mtlalpha0.5_adadelta_bs30_mli800_mlo150//decode_dt_it_beam20_eacc.best_p0_len0.0-0.0_ctcw0.3/result.txt:|       Sum/Avg                      |       1082                79133        |       89.4                 5.8                4.8                 2.9                13.5                98.2        |
exp/tr_it_vggblstmp_e4_subsample1_2_2_1_1_unit320_proj320_d1_unit300_location_aconvc10_aconvf100_mtlalpha0.5_adadelta_bs30_mli800_mlo150//decode_et_it_beam20_eacc.best_p0_len0.0-0.0_ctcw0.3/result.txt:|       SPKR                         |       # Snt               # Wrd        |       Corr                 Sub                Del                 Ins                 Err               S.Err        |
exp/tr_it_vggblstmp_e4_subsample1_2_2_1_1_unit320_proj320_d1_unit300_location_aconvc10_aconvf100_mtlalpha0.5_adadelta_bs30_mli800_mlo150//decode_et_it_beam20_eacc.best_p0_len0.0-0.0_ctcw0.3/result.txt:|       Sum/Avg                      |       1055                77966        |       89.5                 5.7                4.8                 2.7                13.2                98.8        |

```
Currently testing librispeech.

**Comment by sw005320** [comment]
Can you also modify https://github.com/espnet/espnet/blob/master/egs/librispeech/asr1/local/run_spm.sh ?

**Pull request merged**